pkgs/build-support/setup-hooks/desktop-to-darwin-bundle.sh

```diff
@@ -150,7 +150,16 @@ convertDesktopFile() {
     local -r file=$1
     local -r sharePath=$(dirname "$(dirname "$file")")
     local -r name=$(getDesktopParam "${file}" "^Name")
-    local -r exec=$(getDesktopParam "${file}" "Exec")
+    local -r macOSExec=$(getDesktopParam "${file}" "X-macOS-Exec")
+    if [[ "$macOSExec" ]]; then
+      local -r exec="$macOSExec"
+    else
+      local -r execRaw=$(getDesktopParam "${file}" "Exec")
+      local -r exec="${execRaw// %[fFuUick]}"
+      if [[ "$exec" != "$execRaw" ]]; then
+        echo "desktopToDarwinBundle: Application bundles do not understand desktop entry field codes. Changed '$execRaw' to '$exec'."
+      fi
+    fi
```

Comment:
If you `open Foo.app --args <args>`, the `<args>` will passed to the executable. I had to do that at work with the Microsoft Teams app to disable GPU rendering to save battery (before it got that in the app’s preferences). I also have a script that to open a file in Excel (handy for `.csv` files when you don’t want to open them in Numbers but don’t want to make Excel the default app).

However, like I said before, I’m not sure whether those matter or if translating them is the right thing to do on macOS. According to the spec, `%f` and `%u` are used when opening a file or URL. macOS doesn’t seem to care. When I hacked element-desktop to support for the `element:` URL scheme (to use GitHub to authenticate), I don’t think I bothered to change `%f` to `$1`, but I can’t remember. Seems worth revisiting once support for file associations and URL schemes is added.

Setting those aside, it’s the other three (`%i`, `%c`, and `%k`) I’m more concerned about. Those are static (referring back to information in the .desktop file or the .desktop file itself). My assumption is if an app specifies those in its .desktop file, it expects to get that information on the command-line. See my other comment re: apps that use these flags. There is one in nixpkgs I identified (though it is marked as broken on Darwin). I did a search of GitHub, and it seems to be used by Qt apps.

Reply:
I concur that it's better to actually implement the field codes we can.
DEs are supposed to do something smart like spawn a process for every selected file when launching a selection using a desktop entry with `%f`/`%u`. Should we emulate this in the wrapper script? Rather than simply using `$1` we'd have to fork for every argument passed or something.

Reply:
Just pushed some commits handling each of the field codes. (And rebased on master while I was at it.)

I haven't added any special handling of `%f` or `%u`, so multiple arguments passed to such a program would simply silently be ignored. I'm not sure that's very nice behavior. Should we loop over `$@` and fork the program for each argument?

Reply:
> I haven't added any special handling of `%f` or `%u`, so multiple arguments passed to such a program would simply silently be ignored. I'm not sure that's very nice behavior. Should we loop over `$@` and fork the program for each argument?

This is the part I’ve not been sure about, so I did some digging.

First, I put together a simple app that just displays the command-line arguments and associated it with the `public.text` UTI. When I opened text files, it displayed just the `argv[0]`. Not sure whether that was because the app was a SwiftUI app (and the framework was mangling `CommandLine.arguments`), I also tried an Objective-C app that just displays an `NSAlert`. Same thing.

So I did some searching to see how apps are supposed to receive files and URLs. The [answer][so] is Finder calls one of the following methods on the app’s `NSApplicationDelegate`. I don’t know how the invocation works, but if it uses the command-line as part of the process, those arguments do not make it to `main`.

* [`application:openFile`][openFile]
* [`application:openFiles`][openFiles]
* [`application:openURLs`][openURLs]

I think the only approach that will work is to strip `%f` and `%u` and rely on whatever UI framework the app uses to implement the necessary support in its platform-specific stuff for Darwin. Both [GKT][gtk-delegate] and [Qt][qt-delegate] do this, and Electron presumably does as well since my association with the `element:` URL scheme worked without handling it in the script.

[so]: https://stackoverflow.com/questions/2998203/how-to-make-finder-open-with-work-for-my-application-xcode-os-x
[openFile]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428612-application
[openFiles]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428742-application
[openURLs]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/2887193-application
[gtk-delegate]: https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkapplication-quartz.c
[qt-delegate]: https://github.com/qt/qtbase/blob/3fcdb6cb6e35a37f9b511ec2705336102c194d6b/src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm

Reply:
Does this mean I should drop the changes for `%[fFuU]` and strip them out like before? Or are `$1` and `$@` still fine?

Reply:
Well, I doubt it would hurt to keep those changes. If someone `open`ed the app, the file or URL would get passed to it. I’ll go ahead and improve. Thanks for the changes!